src/kernels/MIOpenReduceSum.cpp

```diff
@@ -0,0 +1,141 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+
+__device__ FLOAT_ACCUM warp_reduce_sum(FLOAT_ACCUM val)
+{
+    if(warpSize >= 64)
+        val += __shfl_down(val, 32);
+    if(warpSize >= 32)
+        val += __shfl_down(val, 16);
+    if(warpSize >= 16)
+        val += __shfl_down(val, 8);
+    if(warpSize >= 8)
+        val += __shfl_down(val, 4);
+    if(warpSize >= 4)
+        val += __shfl_down(val, 2);
+    if(warpSize >= 2)
+        val += __shfl_down(val, 1);
+    return val;
+}
+
+__device__ FLOAT_ACCUM block_reduce_sum(FLOAT_ACCUM val)
+{
+    static __shared__ FLOAT_ACCUM shared[REDUCE_SIZE / warpSize];
+    auto lane = threadIdx.x % warpSize;
+    auto wid  = threadIdx.x / warpSize;
+
+    val = warp_reduce_sum(val);
+
+    if(lane == 0)
+        shared[wid] = val;
+    __syncthreads();
+
+    val = threadIdx.x < REDUCE_SIZE / warpSize ? shared[lane] : 0;
+    if(wid == 0)
+        val = warp_reduce_sum(val);
+
+    return val;
+}
+
+template <typename TO>
+__device__ void ReduceSum(const FLOAT_ACCUM* input, TO* output, size_t N)
+{
+    auto gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    FLOAT_ACCUM val = gid < N ? input[gid] : CVT_FP32_2ACCUM(0.0f);
+    if(blockDim.x == warpSize)
+        val = warp_reduce_sum(val);
+    else
+        val = block_reduce_sum(val);
+
+    if(threadIdx.x == 0)
+        output[blockIdx.x] = CVT_ACCUM2FLOAT(val);
+}
+
+extern "C" __global__ void
+ReduceSum(const FLOAT_ACCUM* __restrict__ input, OUTPUT_TYPE* __restrict__ output, size_t N)
+{
+    // instantiate the kernel
+    ReduceSum<OUTPUT_TYPE>(input, output, N);
+}
+
+extern "C" __global__ void ReduceSumFLOATACCUM(const FLOAT_ACCUM* __restrict__ input,
+                                               FLOAT_ACCUM* __restrict__ output,
+                                               size_t N)
+{
+    auto gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    FLOAT_ACCUM val = gid < N ? input[gid] : 0.0f;
+    if(blockDim.x == warpSize)
+        val = warp_reduce_sum(val);
+    else
+        val = block_reduce_sum(val);
+
+    if(threadIdx.x == 0)
+        output[blockIdx.x] = val;
+}
+
+template <typename TO>
+__device__ void Reduce1dSum(const FLOAT_ACCUM* __restrict__ input,
+                            TO* __restrict__ output,
+                            size_t output_numel,
+                            size_t inner_size,
+                            size_t outer_size)
+{
+    int tid  = threadIdx.x;
+    int oidx = blockIdx.x;
+
+    // use double instead of FLOAT_ACCUM for better precision
+    double sum_double = 0.0;
+    for(int i = tid; i < outer_size * inner_size; i += blockDim.x)
+        sum_double +=
+            input[i / inner_size * output_numel * inner_size + oidx * inner_size + i % inner_size];
```

Comment:
Is there a significant difference after changing to double? Please add static_cast<double>.

Reply:
Yes, it is. GTest will failed at least 3 tests if I used FLOAT_ACCUM, which is float in this case. I used to doubt that their were some problems with CPU calculation which lead to floating point precision, but then I'm sure that the CPU code is fine after doing some manual calculation.

Reply:
I have changed type casting to static_cast.

Reply:
Since input is also a FLOAT_ACCUM, static_cast it before adding it to sum_double.

Reply:
I fixed it, so did those in `cpu_prelu.hpp` and `mloPReLUHost.hpp` too.